docs/shaders.rst

```diff
@@ -0,0 +1,325 @@
+Writing PyGfx shaders
+=====================
+
+This document explains how to write shaders for PyGfx. This may be useful
+if you want to improve the existing shaders, add new shaders to PyGfx, or
+if you want to implement custom shaders in your own project.
+
+
+The shader class
+----------------
+
+A shader object derives from ``WorldObjectShader``. Its purposes is to
+provide (templated) shader-code (WGSL), set the corresponding template variables, define
+what bindings (buffers and textures) are used, and provide some more details
+of the pipeline and the rendering.
+
+The shader is associated with a WorldObject-material combination using the ``register_wgpu_render_function()``
+decorator. This decorator can be applied to your shader class, but it can also
+be applied to a function that returns multiple shader objects. This can be useful
+if you want multiple "passes", like a compute pass to prepare some data.
+
+The shader must implement a few methods. A typical shader is shown below:
+
+.. code-block:: python
+
+    from pygfx.renderers.wgpu import (
+        register_wgpu_render_function, WorldObjectShader, Binding, RenderMask
+    )
+
+    @register_wgpu_render_function(SomeWorldObject, SomeMaterial)
+    class SomeShader(WorldObjectShader):
+
+        type = "render"  # must be "render" or "compute"
+
+        def get_resources(self, wobject, shared):
+            bindings = {
+                0: Binding("u_stdinfo", "buffer/uniform", shared.uniform_buffer),
+                ...
+            }
+            self.define_bindings(0, bindings)
+            return {
+                "index_buffer": None,
+                "vertex_buffers": {},
+                "bindings": {
+                    0: bindings,
+                },
+            }
+
+        def get_pipeline_info(self, wobject, shared):
+            return {
+                "primitive_topology": wgpu.PrimitiveTopology.triangle_list,
+                "cull_mode": wgpu.CullMode.none,
+            }
+
+        def get_render_info(self, wobject, shared):
+            n_vertices = ...
+            n_instances = 1
+            render_mask = wobject.render_mask
+            if not render_mask:
+                render_mask = RenderMask.all
```

Comment:
does the renderer default to all if you omit the key, or is it a required key to return?